Feat/course crud

juyeongMoon888 opened the pull request:
## 📘 작업 내용

### Course & Category 도메인 구현
**엔티티**
- `Course`: 강좌의 제목·카테고리·설명·강사·가격·상태·이미지와 생성/수정 시각을 가지며, CourseEdit를 통해 수정됨
- `Category` 강좌를 분류하기 위해 id, 이름, 설명, 노출 순서를 저장하는 엔티티

**DTO**
- `CourseCreateRequest`: 강좌 생성 시 요청 전용 DTO
- `CourseDetailResponse`: 강의 조회용 응답 DTO
- `CategoryResponse`: 카테고리 엔티티에서 id와 이름만 뽑아와, 화면에서 카테고리 목록/선택용으로 사용하는 조회 응답 DTO

**MODEL**
-  DB에 매핑되는 엔티티는 아니면서, 입출력/매핑의 dto도 아님 -> 도메인 내부에서 사용하는 모델로 씀 
- `CourseEdit`: Course 엔티티의 수정 기준으로 사용하는 도메인 모델
- `CourseState`: 강좌 도메인 상태(enum) 값 모델

**Repository**
- 카테고리/강좌 조회 쿼리 메서드

**Service & Exception**
- 강좌 CRUD 비즈니스 로직
- `CourseNotFoundException`: 커스텀 예외 처리

**Test**
- `CourseServiceTest`: 기본 서비스 동작 검증

---

## 🎯 도입 이유

- LXP에서 카테고리와 강좌는 콘텐츠의 "탐색 가능성", "학습자 경험", "운영 효율성"을 
결정짓는 핵심 도메인

---

## ⚙️ 주요 기능

**1. 강의 생성**
```java
Course newCourse = Course.builder()
                .userId(tutorId)
                .categoryId(courseCategoryId)
                .title(courseTitle)
                .description(courseDescription)
                .pointAmount(coursePointAmount)
                .courseState(CourseState.DRAFT)
                .imageUrl(courseImageUrl)
                .build();
```

**2. 조회 기능**
- `getCourse()`: 단일 강좌 조회 
- `getCourses()`: 강좌 리스트 조회
- `getAllCategories()`: 카테고리 리스트 조회

**3. 기술적 특징**
- DTO 변환 `from()` 정적 팩토리 메서드
- @Builder와 스트림을 활용해 강좌·카테고리 조회/응답 로직을 축약하고 중복 코드를 제거
- 응답 DTO 분리로 목록/상세 조회 최적화
- Spring Data JPA 메서드 네이밍 기반 쿼리 자동 생성

---

## ⚠️ 개선 예정

- [ ] 연관관계 매핑
- [ ] 클린코드 기반 리팩터링

juyeongMoon888 committed:
doc: 1차 구현 내용 추가

juyeongMoon888 committed:
feat: Course 생성 추가

juyeongMoon888 committed:
feat: 단일 강좌 조회 서비스/테스트 추가

juyeongMoon888 committed:
feat: 강좌 리스트 조회, 수정, 삭제 서비스/테스트 추가

juyeongMoon888 committed:
feat: 강좌 생성 상태 enum 추가

juyeongMoon888 committed:
feat: 패키지 구조 분리
-dto, domain, controller,repository, service

juyeongMoon888 committed:
feat: course 수정용 DTO 추가

juyeongMoon888 committed:
refactor: @Builder - 빌더 패턴 코드로 수정
- 생성자 제거

juyeongMoon888 committed:
refactor: 캡슐화된 Course → CourseDetailResponse 변환 로직 도입

juyeongMoon888 committed:
refactor: userId, categoryId 수정 필드에서 제거

juyeongMoon888 committed:
refactor: userId, categoryId 수정 필드에서 제거

juyeongMoon888 committed:
refactor: getCourses()-캡슐화&스트림으로 변환

juyeongMoon888 committed:
refactor: 강좌 수정 userId, categoryId 빌드 제거

juyeongMoon888 committed:
refactor: CourseEdit를 강좌수정이라는 도메인 동작으로 위한 전용 객체로 승격

juyeongMoon888 committed:
refactor: Course 엔티티에 강좌 수정 메서드 추가

juyeongMoon888 committed:
refactor: 강좌 수정 로직을 Course 엔티티 메서드로 위임

juyeongMoon888 committed:
feat: CourseNotFound 커스템 예외 추가

juyeongMoon888 committed:
refactor: CourseQueryService -> CourseService 변경

juyeongMoon888 committed:
feat: 카테고리 Entity, Repository, DTO 생성

juyeongMoon888 committed:
feat: 카테고리 조회&선택 서비스 기능 구현, 테스트 작성

juyeongMoon888 committed:
style: 불필요한 import 제거

juyeongMoon888 committed:
refactor: Course 도메인 패키지 구조 정리
- CourseEdit, CourseState를 domain.model로 이동
- entity에는 JPA 엔티티(Course)만 유지

juyeongMoon888 committed:
refactor: Category 도메인 패키지 구조 정리
- Category를 domain.entity 이동

juyeongMoon888 committed:
refactor: Category를 category.domain.entity로 이동 후 import 경로 수정

juyeongMoon888 committed:
refactor: Category를 category.domain.entity로 이동 후 import 경로 수정

juyeongMoon888 committed:
refactor: Category를 category.domain.entity로 이동 후 import 경로 수정

juyeongMoon888 committed:
refactor: 패키지 구조 변경 후 import 경로 수정

juyeongMoon888 committed:
refactor: 패키지 구조 변경 후 import 경로 수정

juyeongMoon888 committed:
style: 불필요한 setter 제거

juyeongMoon888 committed:
style: 불필요한 setter 제거

juyeongMoon888 committed:
feat: course-list.html 템플릿 생성

juyeongMoon888 committed:
refactor: 강좌 조회 DTO에 id와 categoryId 필드 추가

juyeongMoon888 committed:
feat: 강좌·카테고리 목록 조회 컨트롤러 추가

juyeongMoon888 committed:
style: 불필요한 setter 제거

juyeongMoon888 committed:
feat: getId() 추가

juyeongMoon888 committed:
feat: id, categoryName의 getter 추가

juyeongMoon888 committed:
feat: 강사 강좌 관리 화면 및 조회 기능 추가

- /tutor/courses 강좌 관리 뷰 템플릿 추가
- 강사 전용 가좌 목록 조회 컨트롤러 추가
- 강좌 관리 뷰용 읽기 전용 DTO 및 서비스 메서드 추가

dawwson reviewed (commented):
**[ 좋은점 👍 ]**
- 커스텀 exception을 만들어주셔서 어떤 에러인지 명확해지는 효과가 있는 것 같습니다

**[ 개선점 🔄 ]**
- xxxResponse DTO 클래스가 서비스 -> 컨트롤러 사이에서 데이터를 넘기는 목적으로 쓰이고 있는 게 다소 헷갈리는 부분입니다. "Response"라는 이름은 HTTP 요청에 대한 응답 DTO를 연상케 합니다. 따라서 응답 형태로 변환해주는 것은 컨트롤러에서 해주는 것이 좀 더 적절하지 않나... 하는 개인적인 생각입니다! 

**[ 궁금한 점 🤔 ]**
- /model 패키지의 용도는 무엇인가요?? enum과 dto class??가 같이 있어서 여쭤봅니당

dawwson reviewed (approved):
고생하셨습니다!

juyeongMoon888 committed:
refactor: null 무시 규칙으로 부분 수정(PATCH) 지원

sobee1403 commented on src/main/java/com/wanted/growthmate/learning/course/domain/entity/Course.java:
```diff
@@ -0,0 +1,77 @@
+import java.time.LocalDateTime;
+
+@Builder
+@AllArgsConstructor //new Course가 완전히 없어지면 없애도됨
```
생성자 만들어주는 어노테이션인데 요거에 대해 공부하고 사용해주시면 더 좋을 것 같습니다
전 필드를 인자로 받는 생성자가 필요한 케이스가 어디서 발생되는지 찾아서 피드백 주세요

sobee1403 commented on src/main/java/com/wanted/growthmate/learning/course/service/CourseService.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public void deleteCourse(int course_id) {
+        Optional<Course> findCourse = courseRepository.findById(course_id);
```
Optional 에 대해서 공부해보시기~~!

sobee1403 reviewed (commented):
죽여줍니다

juyeongMoon888 committed:
refactor(category): 엔티티 매핑 어노테이션 추가 및 id 자동 생성 설정

- @GeneratedValue(strategy = IDENTITY) 추가로 PK 자동 생성
- 각 필드에 @Column(name = ...) 지정하여 DB 컬럼명 명시
- 매핑 명확화로 JPA 관리 및 유지보수성 향상

juyeongMoon888 committed:
feat: @Column-스키마 제약 추가

sobee1403 reviewed (approved):
굿
